internal/prometheusbp/spectest/spec.go

```diff
@@ -0,0 +1,190 @@
+package spectest
+
+import (
+	"errors"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/testutil"
+
+	"github.com/reddit/baseplate.go/errorsbp"
+)
+
+const (
+	client = "client"
+	server = "server"
+)
+
+var (
+	errPrefix         = errors.New("the prefix is not at the beginning of the metric name")
+	errClientServer   = fmt.Errorf("missing %q or %q as the second part of the metric name", client, server)
+	errLength         = errors.New("metric name should have a minimum of three parts")
+	errNotFound       = errors.New("metric not found")
+	errPrometheusLint = errors.New("prometheus GatherAndLint")
+	errDiffLabels     = errors.New("labels are incorrect")
+)
+
+// ValidateSpec validates that the Prometheus metrics being exposed from baseplate.go
+// conform to the baseplate spec.
+func ValidateSpec(t *testing.T, metricPrefix string, wantMetricCount int) {
+	t.Helper()
+
+	var batch errorsbp.Batch
+	batch.AddPrefix("validate spec", validateSpec(metricPrefix))
+
+	if len(batch.GetErrors()) > 0 {
+		t.Errorf(batch.Compile().Error())
+	}
+}
+
+func validateSpec(prefix string) errorsbp.Batch {
+	var batch errorsbp.Batch
+
+	metricFam, err := prometheus.DefaultGatherer.Gather()
+	if err != nil {
+		return batch
+	}
+
+	metricsUnderTest := buildMetricNames(prefix)
+
+	for _, m := range metricFam {
+		name := *m.Name
+
+		if _, ok := metricsUnderTest[name]; !ok {
+			continue
+		}
+
+		batch.Add(validatePromLint(name))
+		batch.Add(validateName(name, prefix))
+
+		labels := map[string]struct{}{}
+		for _, metric := range m.GetMetric() {
+			for _, label := range metric.Label {
+				labels[*label.Name] = struct{}{}
+			}
+		}
+		batch.Add(validateLabels(name, prefix, labels))
+
+		// after we validate the metric, delete it from the set of all
+		// expected metrics so that we can track if any were not found.
+		delete(metricsUnderTest, name)
+	}
+
+	batch.Add(fmt.Errorf("%w %v", errNotFound, missingMetrics(metricsUnderTest)))
+
+	return batch
+}
+
+func missingMetrics(metrics map[string]struct{}) []string {
```

Comment:
🔕 now that I read it, maybe "keysFrom"?

```go
fmt.Errorf("%w: %v", errNotFound, keysFrom(metricsUnderTest)))
```

Reply:
this will be something like `maps.KeysOf(metricsUnderTest)` after generics I think